#10524 - SetpointManager:ReturnTemperature:Chilled/HotWater (useless) warnings from GetSpecificHeatGlycol

jmarrec opened the pull request:
Pull request overview
---------------------
 - Fixes #10524 

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmarrec committed:
Naive fix for #10524

jmarrec commented on tst/EnergyPlus/unit/SetPointManager.unit.cc:
```diff
@@ -173,9 +174,81 @@ TEST_F(EnergyPlusFixture, SetPointManager_DefineReturnWaterChWSetPointManager)
+
+    compare_err_stream("");
+
+    // Avg temp will be < 0, and before fix it throws this error
+    //   ** Warning ** GetSpecificHeatGlycol: Temperature is out of range (too low) for fluid [WATER] specific heat supplied values **
+    //   **   ~~~   ** ..Called From:ReturnWaterChWSetPointManager::calculate,Temperature=[-1.00], supplied data range=[0.00,125.00]
+    //   **   ~~~   **  Environment=, at Simulation time= 00:00 - 00:00
+    state->dataLoopNodes->Node(1).Temp = 1;
+    state->dataLoopNodes->Node(2).Temp = -3;
+    // Calling it a second time, ensure #10524 is fixed
+    mySPM.calculate(*state, state->dataLoopNodes->Node(1), state->dataLoopNodes->Node(2));
+    // on the first pass through it should detect the plant loop it manages
+    EXPECT_EQ(1, mySPM.plantLoopIndex);
+    // with a delta T of 4, and a target return of 12, it should produce 8
+    EXPECT_EQ(8, mySPM.currentSupplySetPt);
+    compare_err_stream("");
```
Test that shows the issue originally.

jmarrec commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -8282,18 +8279,19 @@ void DefineReturnWaterHWSetPointManager::calculate(EnergyPlusData &state, DataLo
-
     // get the operating flow rate
-    Real64 mdot = supplyNode.MassFlowRate;
+    Real64 const mdot = supplyNode.MassFlowRate;
+    Real64 const deltaT = supplyNode.Temp - returnNode.Temp;
```
That's really all we need

jmarrec commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -8282,18 +8279,19 @@ void DefineReturnWaterHWSetPointManager::calculate(EnergyPlusData &state, DataLo
 
     // check for strange conditions
-    if (Qdemand < 0) {
+    if (deltaT < 0) { // supplyNode.Temp < returnNode.Temp, which shouldn't happen for a HW Loop
```
Checking the sign of Qdemand can be done by checking the sign deltaT only.

jmarrec commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -8321,7 +8319,7 @@ void DefineReturnWaterHWSetPointManager::calculate(EnergyPlusData &state, DataLo
     Real64 T_supply_setpoint = this->maximumHotWaterSetpoint;
     if (mdot > DataConvergParams::PlantFlowRateToler) {
-        T_supply_setpoint = T_return_target + Qdemand / (mdot * cp);
+        T_supply_setpoint = T_return_target + deltaT;
```
Using `Qdemand / (mdot * cp) `means just using `deltaT`

So we do not need the call to GetSpecificHeatGlycol nor the calculation of Qdemand

jmarrec commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -8282,18 +8279,19 @@ void DefineReturnWaterHWSetPointManager::calculate(EnergyPlusData &state, DataLo
 
-    // calculate the current demand
-    Real64 Qdemand = mdot * cp * (supplyNode.Temp - returnNode.Temp);
+    // // calculate the current demand
+    // fluidIndex = state.dataPlnt->PlantLoop(this->plantLoopIndex).FluidIndex;
+    // // we don't need fluid names since we have a real index, so just pass in the temperature and get properties
+    // Real64 const avgTemp = (returnNode.Temp + supplyNode.Temp) / 2;
+    // Real64 const cp = FluidProperties::GetSpecificHeatGlycol(state, "", avgTemp, fluidIndex, "ReturnWaterHWSetPointManager::calculate");
+    // Real64 const Real64 Qdemand = mdot * cp * deltaT;
```
Maybe I should just trash this commented out code?

Myoldmopar commented:
I'm going to hold here.  With @amirroth 's refactor of SPM code, I'm afraid there will be conflicts :) and I'd rather deal with them in this small context PR rather than in that gigantic one.  Please hold and stay on the telephone line, your PR is very important.

Myoldmopar commented:
@jmarrec this will need the conflict(s) resolved.  I haven't looked, but GitHub is offering to let me resolve them online, so they must not be too bad!

jmarrec commented:
@Myoldmopar sure, I'll resolve!

jmarrec committed:
Fix #10524 - The call to GetSpecificGlycol and calculaion of Qdemand is actually pointless

We calculate Qdemand = mdot * cp * deltaT Then we use it as Qdemand / (mdot * cp) so that's the same as just using deltaT only

jmarrec committed:
Add a Unit test for #10524

[ RUN      ] EnergyPlusFixture.SetPointManager_DefineReturnWaterChWSetPointManager_FluidIndex
/Users/julien/Software/Others/EnergyPlus2/tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc:199: Failure
Expected equality of these values:
  expected_string
    Which is: ""
  stream_str
    Which is: "   ** Warning ** GetSpecificHeatGlycol: Temperature is out of range (too low) for fluid [WATER] specific heat supplied values **\n   **   ~~~   ** ..Called From:ReturnWaterChWSetPointManager::calculate,Temperature=[-1.00], supplied data range=[0.00,125.00]\n   **   ~~~   **  Environment=, at Simulation time= 00:00 - 00:00\n"
With diff:
@@ -1,1 +1,3 @@
-""
+   ** Warning ** GetSpecificHeatGlycol: Temperature is out of range (too low) for fluid [WATER] specific heat supplied values **
+   **   ~~~   ** ..Called From:ReturnWaterChWSetPointManager::calculate,Temperature=[-1.00], supplied data range=[0.00,125.00]
+   **   ~~~   **  Environment=, at Simulation time= 00:00 - 00:00\n

jmarrec commented:
Replayed the changes onto develop manually.

Myoldmopar commented on tst/EnergyPlus/unit/SetPointManager.unit.cc:
```diff
@@ -1517,7 +1590,8 @@ TEST_F(EnergyPlusFixture, SingZoneCoolHeatSetPtMgrZoneInletNodeTest)
         "   ..... Reference severe error count=2",
-        "   ..... Last severe error=SetpointManager:SingleZone:Cooling=\"COOLING SUPPLY AIR TEMP MANAGER 1\", The zone inlet node of ZNF1 INLET NODE",
+        "   ..... Last severe error=SetpointManager:SingleZone:Cooling=\"COOLING SUPPLY AIR TEMP MANAGER 1\", The zone inlet node of ZNF1 INLET "
+        "NODE",
```
FWIW, some of these error snippets that have to match _exactly_ could be modified to use [`compare_err_stream_substring`](https://github.com/NREL/EnergyPlus/blob/92ba541ca525940013c228b84b7644d31212118f/tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.hh#L193).  That way you can just make sure it has a reasonable message, and not have these full text blocks that have to match exactly.

I know you didn't do a functional change here, it just line wrapped, but I'm saying it anyway :)

Myoldmopar reviewed (approved):
This seems quite fine to me.

Myoldmopar commented:
I think we'll get this merged now, thanks @jmarrec
